Implement download button feature for logs

TheRealJon opened the pull request:
# Description
Log content can be downloaded via a download button on the top right corner of the log window. This downloads the currently displayed content (up to the last 1000 lines of the log).

# Changes
- Refactored `co-m-pane__top-controls` to `top-controls` and now uses flexbox.
  - `top-controls__group` and modifiers `top-controls__group--right`, and `--left` can now be used as children to distribute control item groups across the top controls element. Items within these groups are aligned as the class name suggests (center by default). This is a wrapper for a group of control items in the top control element.
  - `top-controls__item` should be used as a child of a top controls group element and are automatically vertically centered, and spaced 15px apart horizontally. This is a wrapper for each control item in a top controls group element.
- Implemented download functionality in the ResourceLog component
- Removed obsolete `_resource-log.scss`. Top controls style refactoring made these styles unnecessary.
- Update PodExec component to use new top controls styles (only other place where top controls was used).

# Screenshots
![screenshot-localhost-9000-2018 05 29-12-59-18](https://user-images.githubusercontent.com/22625502/40673641-5079d5d2-6340-11e8-934e-7561235b6945.png)

320px screen width
![screenshot-localhost-9000-2018 05 30-12-15-44](https://user-images.githubusercontent.com/22625502/40733848-b2345076-6404-11e8-8abb-277ede557c92.png)


@openshift/team-ux-review

rhamilto commented on frontend/public/components/pod-exec.jsx:
```diff
@@ -144,10 +144,14 @@ export class PodExec extends React.PureComponent {
-          Connecting to
-        </span>
-        <Dropdown className="btn-group" items={_.mapValues(containers, nameWithIcon)} title={nameWithIcon(activeContainer || <LoadingInline />)} onChange={this.onChangeContainer} />
+        <div className="co-m-pane__top-controls__group--left">
```
BEM suggests you shouldn't nest elements.  http://getbem.com/faq/#css-nested-elements

ggreer commented:
It really feels like this should be two pull requests: One to add the download button, and one to refactor layout/styling.

TheRealJon commented:
@rhamilto I changed the `co-m-pane__top-controls` class to just `top-controls` and adjusted the child classes as well. The controls also reflow at mobile widths so that the controls on the left stack. Posted an updated screen shot at 320px width.

rhamilto commented:
Nice improvement, @TheRealJon!  

I find it a bit odd that "Log streaming..." no longer sits next to the pause button.  Is it possible to maintain the button and text on one line?

spadgett commented on frontend/public/style/_common.scss:
```diff
@@ -85,8 +85,48 @@
+      justify-content: flex-end;
+    }
+
+    .top-controls__item {
```
I suggest we flatten this. We shouldn't need the nesting with BEM.

TheRealJon commented:
Just updated the styles/structure of the top controls for the ResourceLog component. The message next to the TogglePlay button will not wrap @ mobile anymore. Also flattened the scss classes in `_common.scss`.

@rhamilto @spadgett @beanh66

beanh66 commented:
Sweet, thanks @TheRealJon LGTM!

rhamilto commented on frontend/public/components/_toggle-play.scss:
```diff
@@ -6,6 +6,10 @@
   width: 32px;
   padding: 0;
 
+  &--space-r {
```
Nit:  I'm not a big fan of class names that describe what they do.  I would rather see a class name that semantically describes the content rather than the presentation.  Additionally, you could just put the margin right on the parent as the other context that uses ToggleButton [has an override that sets the margin to 0](https://github.com/openshift/console/blob/master/frontend/public/components/_sysevent-stream.scss#L60).

TheRealJon replied:
Yeah, I was trying to avoid adding a new class just for this context since this pattern might be used elsewhere. One option would be to make the status message an optional property of the TogglePlay component itself so that we can include styling in that component.

ggreer commented:
Matt and I spent close to an hour trying to figure out the implications of this markup and styling. The combination of flexbox, media queries, nested elements, and padding tweaks... it all made us kinda give up trying to understand the exact differences regarding what users would see. I also wondered whether bootstrap or patternfly already had the necessary functionality. If not, perhaps we should choose a different UI pattern.

That said: The code behind the download functionality looks good, and this is a desirable feature. Had it been in its own PR with sub-optimal styling, I'm sure it would have been merged already. But when these two changes are coupled in the same PR, it tends to cause people to want to merge prematurely. (Since the sooner it's merged, the sooner the desirable feature ships.) I want to avoid that.

ggreer commented:
jenkins rebuild

spadgett commented:
@ggreer Do you have specific suggestions for how to make this simpler?

We were looking at it this morning, and flexbox seems reasonable. I'm not sure we can achieve the same layout with say Bootstrap grid. The intent is to make top-controls reusable for other pages so we only solve this once.

spadgett commented:
`pull-right` doesn't work well because the vertical alignment is off, which flexbox fixes

TheRealJon commented:
squashed

ggreer commented:
I think it's fine to float the button and tweak the styling in a separate PR. Users won't drop a product because a button is misaligned by a few pixels. What I'm more worried about is our markup and styling growing in complexity, slowing development speed and increasing the likelihood of regressions.

This PR is an excellent example of what I'm talking about. Here's the entirety of the log download functionality:

```javascript
import { saveAs } from 'file-saver';
...
const LogControls = ({dropdown, onClickDownload, status, toggleStreaming}) => {
...
        <button className="btn btn-default" onClick={onClickDownload}>
          <i className="fa fa-download"></i>&nbsp;Download
        </button>
...
this._download = this._download.bind(this);
...
  _download(){
    const blob = new Blob([this._buffer.lines().join('')], {type: 'text/plain;charset=utf-8'});
    const filename = this.props.containerName || this.props.resourceName;
    saveAs(blob, `${filename}.log`);
  }
...
onClickDownload={this._download}
```

That's 12 lines. Do we really need another 93 lines of changes to fix a few pixels of vertical offset? I don't think so.

spadgett commented:
It's not just for vertical alignment. We want to reflow for mobile as well.

But I just don't agree that this is that complicated. Please offer an alternative if there is an easier way. I'm always happy if we can simplify things and still achieve the layout we want.

ggreer commented:
Here's a commit that gives the same behavior: https://github.com/openshift/console/commit/fbb65d276d4f24a9939a488505d9df7057121eda

spadgett commented:
Thanks @ggreer. Appreciate that.

It's less code, but I have to point out that it doesn't reflow well, which is one of the problems we wanted to solve.

![rook-operator-84f45dc8b4-fnw6r details tectonic 2018-06-01 12-22-29](https://user-images.githubusercontent.com/1167259/40851829-8553378c-6596-11e8-9c74-600f21663f0d.png)

![rook-operator-84f45dc8b4-fnw6r details tectonic 2018-06-01 12-23-21](https://user-images.githubusercontent.com/1167259/40851869-98182116-6596-11e8-9f9d-fb698405d2fe.png)

Also the "Log streaming..." text isn't vertically centered:

![rook-operator-84f45dc8b4-fnw6r details tectonic 2018-06-01 12-25-46](https://user-images.githubusercontent.com/1167259/40851976-f5b3ee54-6596-11e8-9082-1712cd68d824.png)

ggreer commented:
Oops, here's a revised commit with the log streaming text centered: https://github.com/openshift/console/commit/bebb60063127e4dca7cb28bf5ce1b0e75d470e04 I pushed an image built with this commit (quay.io/coreos/tectonic-console:bebb60063127e4dca7cb28bf5ce1b0e75d470e04) and deployed it to teamui.

I realize the mobile styling isn't great, but log view on mobile has much bigger problems than the padding between buttons. For example, the viewable log area is barely larger than a postage stamp.

beanh66 commented:
> I am wondering if it's better to have that "Download" btn being styled as a btn-link (with download icon)?

@tlwu2013 We had this same debate while you were out and I believe ended up with a button to be consistent with Tectonic for now, but intended to revisit in the near future when we add other functionality like expand. Not sure we realized there was already some precedent for using links like this in Tectonic though. Having seen the screens you reference, I agree we could just go with a link from the start! @serenamarie125 @cshinn Thoughts?

spadgett commented:
@TheRealJon Can you take @ggreer's latest commit for the styling? The only issue I see is that there's no margin between the items when they stack, but I think that shouldn't be too difficult to fix (either in this PR or a follow on).

@tlwu2013 I like using a link. If we change it here, should we change the download button on the YAML page, too?

I notice the case is different in "Reveal Values" compared to the others ("View samples"). I think we should fix that for consistency.

TheRealJon commented:
@spadgett @ggreer Switched over to the simplified styles @ggreer suggested.

spadgett commented on frontend/public/components/utils/resource-log.jsx:
```diff
@@ -22,14 +23,17 @@ const dataHasHTML = (data) => {
       {logStatusMessages[status]}
     </span>
-    {dropdown && <span>{dropdown}</span>}
+    {dropdown && <span style={{flexGrow: 1}}>{dropdown}</span>}
```
We should avoid inline styles. (I know this was in @ggreer's commit.)

spadgett commented on frontend/public/components/utils/resource-log.jsx:
```diff
@@ -22,14 +23,17 @@ const dataHasHTML = (data) => {
-    {dropdown && <span>{dropdown}</span>}
+    {dropdown && <span style={{flexGrow: 1}}>{dropdown}</span>}
+    <button className="btn btn-default" onClick={onDownload}>
+      <i className="fa fa-download"></i>&nbsp;Download
```
aria-hidden="true"

TheRealJon commented:
Once we get some input from @openshift/team-ux-review about whether the button should be `btn-link` or `btn-default`, this should be ready to go.

stevekuznetsov commented:
Merge blocking on the branch for this repo is causing our merge infrastructure to crash on this PR:

```
{"base-sha":"6805cee966d08b09a8ceee795016b060a6ec6351","branch":"master","component":"tide","controller":"sync","error":"1 review requesting changes and 2 approving reviews by reviewers with write access.","level":"error","msg":"Merge failed: PR is unmergable. How did it pass tests?!","org":"openshift","pr":77,"repo":"console","sha":"2fbd13414a8703c0590be130c8909c1620c14a6b","time":"2018-06-04T16:29:59Z"}
```

If you want merge automation we need to allow the bot to merge.

stevekuznetsov commented:
/hold

(so the bot leaves it alone)

spadgett commented:
/hold cancel
